plugins/inputs/execd/execd.go

```diff
@@ -0,0 +1,209 @@
+package execd
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"os/exec"
+	"sync"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"github.com/influxdata/telegraf/plugins/parsers"
+)
+
+const sampleConfig = `
+  ## Program to run as daemon
+  command = ["telegraf-smartctl", "-d", "/dev/sda"]
+
+  ## Define how the process is signaled on each collection interval.
+
+  ## Valid values are:
+  ##   "none"   : Do not signal anything.
+  ##              The process must output metrics by itself.
+  ##   "STDIN"  : Send a newline on STDIN.
+  ##   "SIGHUP" : Send a HUP signal. Not available on Windows.
+  signal = "none"
+
+  ## Delay before the process is restarted after an unexpected termination
+  restart_delay = "10s"
+
+  ## Data format to consume.
+  ## Each data format has its own unique set of configuration options, read
+  ## more about them here:
+  ## https://github.com/influxdata/telegraf/blob/master/docs/DATA_FORMATS_INPUT.md
+  data_format = "influx"
+`
+
+type Execd struct {
+	Command      []string
+	Signal       string
+	RestartDelay internal.Duration
+
+	acc    telegraf.Accumulator
+	cmd    *exec.Cmd
+	parser parsers.Parser
+	stdin  io.WriteCloser
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+}
+
+func (e *Execd) SampleConfig() string {
+	return sampleConfig
+}
+
+func (e *Execd) Description() string {
+	return "Run executable as long-running input plugin"
+}
+
+func (e *Execd) SetParser(parser parsers.Parser) {
+	e.parser = parser
+}
+
+func (e *Execd) Start(acc telegraf.Accumulator) error {
+	e.acc = acc
+
+	if len(e.Command) == 0 {
+		return fmt.Errorf("E! [inputs.execd] FATAL no command specified")
+	}
+
+	e.wg.Add(1)
+
+	var ctx context.Context
+	ctx, e.cancel = context.WithCancel(context.Background())
+
+	go func() {
+		e.cmdLoop(ctx)
+		e.wg.Done()
+	}()
+
+	return nil
+}
+
+func (e *Execd) Stop() {
+	e.cancel()
+	e.wg.Wait()
+}
+
+func (e *Execd) cmdLoop(ctx context.Context) {
+	for {
+		// Use a buffered channel to ensure goroutine below can exit
+		// if `ctx.Done` is selected and nothing reads on `done` anymore
+		done := make(chan error, 1)
+		go func() {
+			done <- e.cmdRun()
+		}()
+
+		select {
+		case <-ctx.Done():
+			// Immediately exit process but with a graceful shutdown
+			// period before killing
+			internal.WaitTimeout(e.cmd, 0)
```

Comment:
this isn't actually stopping the process. We need to send it a signal and close stdin.